cli/state.rs

```diff
@@ -195,9 +196,14 @@ impl ThreadSafeState {
       Some(file_name) => {
         let base_url = match &main_module {
           Some(module_specifier) => module_specifier.clone(),
-          None => unreachable!(),
+          // Dummy child of the CWD.
+          None => ModuleSpecifier::resolve_url_or_path("_").unwrap(),
```

Comment:
It's not dummy - `main_module` would then actually be file `_` in cwd, it's not right. It should be just cwd

Reply:
Yes, this is an ugly workaround because `base_url` is taken as a sibling, not as a parent. Either import_map.rs or `Url::join()` needs to be refactored so it takes a parent (the `_` would change to a `.` in that case).

I found that `Url::join()` also works desirably with the CWD with a slash at the end:
```rust
let base_url = match &main_module {
  Some(module_specifier) => module_specifier.to_string(),
  None => {
    format!("file://{}/", env::current_dir().unwrap().to_str().unwrap())
  }
};
let import_map = ImportMap::load(&base_url, file_name)?;
```
Would this have been better?

Reply:
@nayeem-rahman yes, that should work, I'd suggest going with:
```
let base_url = match &main_module {
  Some(module_specifier) => module_specifier.to_string(),
  None => {
    Url::from_file_path(env::current_dir().unwrap()).unwrap()
  }
};
```

Reply:
It's not working correctly. Like I said, the import map root is set to the parent of whatever `base_url` you give so it works properly when you give `main_module`. If you want to make the CWD the root, you have to give some child of the CWD (or you can put a slash at the end like I snippeted, but there is no suitable API that produces this naturally).

Reply:
You can test it like this:
./a.js:
```js
import("mappedToB").then(console.log).catch(console.error);
```
./b.js:
```js
export const data = 10;
```
./importmap.json:
```json
{
  "imports": {
    "mappedToB": "./b.js"
  }
}
```
Run `cat a.js | target/debug/deno --importmap importmap.json`.
If it works you'll see `{ data: 10 }`, otherwise you'll see an error saying that b.js couldn't  be found and that it was searching in the parent directory.

Reply:
Alright, I'm fine with your solution. My only concern is it will work on Windows properly, but we'll see on CI :)

Reply:
@bartlomieju Wait a minute. Shouldn't the import map root be the import map's directory? It's currently being set to to the _main module_'s directory (or newly the CWD if there isn't one), but it should be the import map's location in both cases according to the spec. Was this a mistake?

Reply:
@nayeem-rahman I don't think that's true, could you link relevant part of spec?

Reply:
@bartlomieju https://github.com/WICG/import-maps#bare-specifiers-for-javascript-modules
> In the case of relative-URL-like addresses, they are resolved relative to the import map's base URL, i.e. the base URL of the page for inline import maps, and the URL of the import map resource for external import maps.

Reply:
Hm that's valid then :+1: not sure how I missed that... Well since we're using URL to specify import map I guess we should treat it as external import map.

Reply:
In light of this, I've moved the base URL resolution inside `ImportMap::load()` and the API is significantly simpler: https://github.com/denoland/deno/pull/2770/commits/b055ac276fa848880f8f39326c2b0e48ce3b0f91